# Лабораторная работа № 2

Группа М8О-109СВ-24

Панкин Андрей Андреевич

(Предварительный вывод: Мне не понравилось, что для pygmo пришлось устанавливать conda)

---


In [6]:
import pygmo as pg
import numpy as np
import pandas as pd

# Тестовые функции #

In [7]:
class ThreeHumpCamelFunction:
    def fitness(self, x):
        return [2 * x[0]**2 - 1.05 * x[0]**4 + (x[0]**6) / 6 + x[0] * x[1] + x[1]**2]  # Трёхгорбый верблюд

    def get_bounds(self):
        return ([-5, -5], [5, 5])  

class McCormickFunction:
    def fitness(self, x):
        return [np.sin(x[0] + x[1]) + (x[0] - x[1])**2 - 1.5 * x[0] + 2.5 * x[1] + 1]  # МакКормик

    def get_bounds(self):
        return ([-1.5, -3], [4, 4]) 


In [ ]:
functions = [
    ("ThreeHumpCamel", ThreeHumpCamelFunction()),
    ("McCormick", McCormickFunction())
]

---

# Алгоритмы нахождения оптимума #

In [8]:
algorithms = [
    ("DE", pg.algorithm(pg.de())),  # Дифференциальная эволюция
    ("PSO", pg.algorithm(pg.pso())),  # Рой частиц
    ("SA", pg.algorithm(pg.simulated_annealing()))  # Имитация отжига
]

In [9]:
results = []

for func_name, func in functions:
    prob = pg.problem(func)
    for algo_name, algo in algorithms:
        solutions = []
        for _ in range(5):
            archi = pg.archipelago(n=1, algo=algo, prob=prob, pop_size=20)
            archi.evolve()
            archi.wait()
            sols = archi.get_champions_f()
            solutions.extend(sols)
        avg_fitness = np.mean(solutions)
        results.append({"Function": func_name, "Algorithm": algo_name, "Average Fitness": avg_fitness})

results_df = pd.DataFrame(results)

In [10]:
print(results_df)

         Function Algorithm  Average Fitness
0  ThreeHumpCamel        DE         0.367023
1  ThreeHumpCamel       PSO         0.345507
2  ThreeHumpCamel        SA         0.005550
3       McCormick        DE        -1.567059
4       McCormick       PSO        -1.654509
5       McCormick        SA        -1.897929


---

В ходе выполнения лабораторной работы были протестированы две тестовые функции оптимизации:  
- **Функция трёхгорбого верблюда**  
- **Функция МакКормика**  

Для их оптимизации применялись три алгоритма из библиотеки Pygmo:  
- Дифференциальная эволюция (**DE**)  
- Рой частиц (**PSO**)  
- Имитация отжига (**SA**)  

Результаты экспериментов представлены в таблице ниже:

| Function         | Algorithm | Average Fitness |
|------------------|-----------|-----------------|
| ThreeHumpCamel   | DE        | 0.367023        |
| ThreeHumpCamel   | PSO       | 0.345507        |
| ThreeHumpCamel   | SA        | 0.005550        |
| McCormick        | DE        | -1.567059       |
| McCormick        | PSO       | -1.654509       |
| McCormick        | SA        | -1.897929       |

### Анализ результатов:
1. На функции **трёхгорбого верблюда** лучшего результата достиг метод **SA (имитация отжига)**, показав минимальное значение средней целевой функции (0.005550).  
2. На функции **МакКормика** метод **SA** также продемонстрировал лучший результат, с минимальным значением средней целевой функции (-1.897929).  
3. Методы **DE** и **PSO** показали схожие результаты, но уступили **SA** на обеих функциях.  

### Выводы:
1. Метод имитации отжига (**SA**) показал себя наиболее эффективным на задачах оптимизации функций с несколькими экстремумами.  
2. Методы дифференциальной эволюции (**DE**) и роя частиц (**PSO**) также продемонстрировали хорошие результаты, но они могут быть менее точными на некоторых задачах.  
3. Выбор подходящего алгоритма зависит от особенностей задачи и требуемой точности.  

Работа с библиотекой Pygmo мне не очень понравилась в силу того, что она не устанавливается через pip.
